### ISSUE TITLE
Crash in validation of transport params containing "0.0.0.0"

### ISSUE THREAD

**Issue by Paul-calrec** [sony/nmos-cpp#182]
Hi All
I'm getting a crash in a situation where the secondary leg of the SDP contains:
m=audio 0 RTP/AVP 97\r\nc=IN IP4 0.0.0.0/32\r\na=mediaclk:direct=0\r\na=ptime:1\r\na=ts-refclk:ptp=IEEE1588-2008:00-90-56-FF-FE-08-0A-B1:0\r\na=source-filter: incl IN IP4 0.0.0.0 0.0.0.0\r\na=rtpmap:97 L24/48000/8\r\na=fmtp:97 channel-order=SMPTE2110.(U08)\r\na=mid:secondary\r\n

As you can see all of the IP addresses are 0.0.0.0 - it is not obvious to me whether this is valid SDP but it is from a third party device so not so easy for me to fix. The primary leg of the SDP is fine. 

When our receiver's connection API is given this SDP it crashes in nlohmann/json-validator.cpp, validate(), line 281 due to calling s->validate where s is null.

I'm fairly new to this code so any clues would be appreciated. Should the code be able to handle this SDP?
Regards

Paul Jackson

**Comment by garethsb** [comment]
Hi Paul, I haven't been immediately able to reproduce this. Can you supply more of the call stack, the whole `transport_file`.`data` contents and the value of the `/constraints` endpoint for the Receiver in question?

**Comment by lo-simon** [comment]
Hi Paul,

To verify your SDP you can try out sdpoker, which can be found at https://github.com/Streampunk/sdpoker.

Can you share your entity SDP with us? That could help us easier to identify the issue.

**Comment by Paul-calrec** [comment]
Hi Gareth, Simon, we are a few months behind master, we are on commit ae3356c2c7ee7213ffe439c0ac643a2fad4bc5c1 from 11th May 2020.

This is what we are sending to the receiver staged uri to reproduce the issue:
{ 
 "master_enable": true, 
 "transport_file":  { 
  "data": "v=0\r\no=- 2 1479834355 IN IP4 192.168.30.41\r\ns=912/UCP/051 Spigot 2\r\ni=UCP NMOS client\r\nt=0 0\r\na=sendonly\r\na=group:DUP primary secondary\r\nm=audio 5004 RTP/AVP 97\r\nc=IN IP4 238.30.51.2/32\r\na=mediaclk:direct=0\r\na=ptime:1\r\na=ts-refclk:ptp=IEEE1588-2008:00-90-56-FF-FE-08-0A-B1:0\r\na=source-filter: incl IN IP4 238.30.51.2 192.168.30.111\r\na=rtpmap:97 L24/48000/8\r\na=fmtp:97 channel-order=SMPTE2110.(U08)\r\na=mid:primary\r\nm=audio 0 RTP/AVP 97\r\nc=IN IP4 0.0.0.0/32\r\na=mediaclk:direct=0\r\na=ptime:1\r\na=ts-refclk:ptp=IEEE1588-2008:00-90-56-FF-FE-08-0A-B1:0\r\na=source-filter: incl IN IP4 0.0.0.0 0.0.0.0\r\na=rtpmap:97 L24/48000/8\r\na=fmtp:97 channel-order=SMPTE2110.(U08)\r\na=mid:secondary\r\n", 
  "type": "application/sdp" 
 }, 
 "activation":  { 
  "mode": "activate_immediate" 
 }
}

The constraints are:
"constraints": [
    {
      "destination_port": {},
      "interface_ip": "192.168.30.111",
      "multicast_ip": {},
      "rtp_enabled": {},
      "source_ip": {}
    },
    {
      "destination_port": {},
      "interface_ip": "192.168.31.111",
      "multicast_ip": {},
      "rtp_enabled": {},
      "source_ip": {}
    }
  ]

Here is the call stack:
Crash reason:  SIGSEGV
Crash address: 0x0
Process uptime: not available

Thread 47 (crashed)
 0  libnmos_protocol_plugin.so!logical_combination<(<unnamed>::logical_combination_types)1>::validate [json-validator.cpp : 281 + 0x4]
     r0 = 0x00000000    r1 = 0x9cdaaf48    r2 = 0x954ef210    r3 = 0x9cdaadd8
     r4 = 0xb4a111c8    r5 = 0x9cdaae08    r6 = 0x9cdaadd8    r7 = 0x9cdaadc8
     r8 = 0x9cdaadc0    r9 = 0x95479590   r10 = 0x9cdaade0   r12 = 0x95479370
     fp = 0x95479548    sp = 0x9cdaada8    lr = 0xb1ebc5d4    pc = 0xb1ebc5ec
    Found by: given as instruction pointer in context
 1  libnmos_protocol_plugin.so!type_schema::validate [json-validator.cpp : 384 + 0x16]
     r4 = 0x9540fb70    r5 = 0x954c13c8    r6 = 0x954ef210    r7 = 0x9cdaaf48
     r8 = 0x954ed2e8    r9 = 0x9cdab524   r10 = 0x9540fb78    fp = 0xb1ebc500
     sp = 0x9cdaae40    pc = 0xb1ebae3c
    Found by: call frame info
 2  libnmos_protocol_plugin.so!object::validate [json-validator.cpp : 784 + 0x12]
     r4 = 0x9cdaaf60    r5 = 0x00000000    r6 = 0x9cdaaf78    r7 = 0x9cdaaf48
     r8 = 0x9cdaaf68    r9 = 0x954ed2e8   r10 = 0xb1ebac68    fp = 0x0000000c
     sp = 0x9cdaaec0    pc = 0xb1ebe73c
    Found by: call frame info
 3  libnmos_protocol_plugin.so!type_schema::validate [json-validator.cpp : 363 + 0x16]
     r4 = 0xb1ebe318    r5 = 0xb1fb1000    r6 = 0x954e40c8    r7 = 0x9cdab0e8
     r8 = 0x954ec6e8    r9 = 0x9cdab524   r10 = 0x9cdab0f4    fp = 0x9cdab110
     sp = 0x9cdab038    pc = 0xb1ebace4
    Found by: call frame info
 4  libnmos_protocol_plugin.so!array::validate [json-validator.cpp : 932 + 0x16]
     r4 = 0x9cdab100    r5 = 0x9cdab0e8    r6 = 0xb4a33ce4    r7 = 0x954ec6e8
     r8 = 0x9cdab1b0    r9 = 0x95466348   r10 = 0x9cdab0f4    fp = 0x9cdab110
     sp = 0x9cdab0b8    pc = 0xb1ebc198
    Found by: call frame info
 5  libnmos_protocol_plugin.so!type_schema::validate [json-validator.cpp : 363 + 0x16]
     r4 = 0xb1ebbae8    r5 = 0xb1fb1000    r6 = 0x954595b0    r7 = 0x9cdab328
     r8 = 0xb4a30d38    r9 = 0x9cdab524   r10 = 0xb1ebac68    fp = 0x00000010
     sp = 0x9cdab220    pc = 0xb1ebace4
    Found by: call frame info
 6  libnmos_protocol_plugin.so!object::validate [json-validator.cpp : 784 + 0x12]
     r4 = 0x9cdab340    r5 = 0x00000000    r6 = 0x9cdab358    r7 = 0x9cdab328
     r8 = 0x9cdab348    r9 = 0xb4a30d38   r10 = 0xb1ebac68    fp = 0x00000010
     sp = 0x9cdab2a0    pc = 0xb1ebe73c
    Found by: call frame info
 7  libnmos_protocol_plugin.so!type_schema::validate [json-validator.cpp : 363 + 0x16]
     r4 = 0xb1ebe318    r5 = 0xb1fb1000    r6 = 0x9cdab530    r7 = 0x9cdab4cc
     r8 = 0xb4a7a078    r9 = 0x9cdab524   r10 = 0x0035c0f4    fp = 0x9cdab5e8
     sp = 0x9cdab418    pc = 0xb1ebace4
    Found by: call frame info
 8  libnmos_protocol_plugin.so!schema_ref::validate [json-validator.cpp : 58 + 0xe]
     r4 = 0xb1ebac68    r5 = 0xb1eba650    r6 = 0x9cdab4f0    r7 = 0x9cdab524
     r8 = 0x9cdab530    r9 = 0x9441172c   r10 = 0x0035c0f4    fp = 0x9cdab5e8
     sp = 0x9cdab498    pc = 0xb1eba674
    Found by: call frame info
 9  libnmos_protocol_plugin.so!nlohmann::json_schema::json_validator::validate(nlohmann::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long long, unsigned long long, double, std::allocator, nlohmann::adl_serializer> const&, nlohmann::json_schema::error_handler&) const [json-validator.cpp : 208 + 0x16]
     r4 = 0x9cdab4cc    r5 = 0xb1eba650    r6 = 0x9cdab4f0    r7 = 0x9cdab524
     r8 = 0x9cdab530    r9 = 0x9441172c   r10 = 0x0035c0f4    fp = 0x9cdab5e8
     sp = 0x9cdab4c8    pc = 0xb1ebb9ac
    Found by: call frame info
10  libnmos_protocol_plugin.so!web::json::experimental::details::json_validator_impl::validate(web::json::value const&, web::uri const&) const [json_validator_impl.cpp : 125 + 0xe]
     r4 = 0x9cdab530    r5 = 0x9cdab578    r6 = 0x94411670    r7 = 0x9cdab528
     r8 = 0xb1fb1000    r9 = 0x9cdab540   r10 = 0x0035c0f4    fp = 0x9cdab5e8
     sp = 0x9cdab520    pc = 0xb1e14890
    Found by: call frame info
11  libnmos_protocol_plugin.so!nmos::details::validate_staged_constraints(nmos::type const&, web::json::value const&, nmos::transport const&, web::json::value const&) [connection_api.cpp : 301 + 0xe]
     r4 = 0x0035c18c    r5 = 0x9cdab5c0    r6 = 0x9cdab5d8    r7 = 0x0035c0f4
     r8 = 0x0035c048    r9 = 0x0035c0c8   r10 = 0x0035c0f4    fp = 0x9cdab5e8
     sp = 0x9cdab5a8    pc = 0xb1d4dfac
    Found by: call frame info
12  libnmos_protocol_plugin.so!nmos::details::handle_connection_resource_patch(nmos::node_model&, std::unique_lock<boost::shared_mutex>&, nmos::api_version const&, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, nmos::type> const&, web::json::value const&, nmos::tai const&, std::function<web::json::value (nmos::resource const&, nmos::resource const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, slog::base_gate&)>, std::function<void (nmos::resource const&, nmos::resource const&, web::json::value const&, slog::base_gate&)>, slog::base_gate&) [connection_api.cpp : 740 + 0x12]
     r4 = 0x9cdab7f0    r5 = 0x9cdab7f8    r6 = 0x9cdab790    r7 = 0x9cdab938
     r8 = 0x9cdab7c0    r9 = 0xaf48d650   r10 = 0xaf48d720    fp = 0x9cdaba24
     sp = 0x9cdab768    pc = 0xb1d4f3b0
    Found by: call frame info
13  libnmos_protocol_plugin.so!nmos::details::handle_connection_resource_patch(web::http::http_response, nmos::node_model&, nmos::api_version const&, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, nmos::type> const&, web::json::value const&, std::function<web::json::value (nmos::resource const&, nmos::resource const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, slog::base_gate&)>, std::function<void (nmos::resource const&, nmos::resource const&, web::json::value const&, slog::base_gate&)>, slog::base_gate&) [connection_api.cpp : 792 + 0x3a]
     r4 = 0x9cdab950    r5 = 0x9cdab968    r6 = 0xb2258370    r7 = 0x9cdab930
     r8 = 0x9cdab9d4    r9 = 0x9cdab9dc   r10 = 0x9cdaba24    fp = 0x9cdab9dc
     sp = 0x9cdab908    pc = 0xb1d4feec
    Found by: call frame info
14  libnmos_protocol_plugin.so!nmos::make_unmounted_connection_api(nmos::node_model&, std::function<web::json::value (nmos::resource const&, nmos::resource const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, slog::base_gate&)>, std::function<void (nmos::resource const&, nmos::resource const&, web::json::value const&, slog::base_gate&)>, slog::base_gate&)::{lambda(web::http::http_request, web::http::http_response, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&)#9}::operator()(web::http::http_request, web::http::http_response, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) const::{lambda(web::json::value)#1}::operator()(web::json::value) + 0x1be
     r4 = 0x9cdaba54    r5 = 0x9cdab9e4    r6 = 0x95244e60    r7 = 0x9cdaba24
     r8 = 0x9cdaba14    r9 = 0x9cdab9dc   r10 = 0x9cdab9d4    fp = 0x9cdab9fc
     sp = 0x9cdab9b0    pc = 0xb1d50350
    Found by: call frame info

Thanks, Paul

**Comment by garethsb** [comment]
Hi Paul,

It looks to me like your constraints aren't in quite the right format, rather than
```json
"interface_ip": "192.168.30.111",
```
you need
```json
"interface_ip": { "enum": ["192.168.30.111"] },
```

There's a helper function if you want it.

That'll probably then cause the patch to be rejected, as the `enum` doesn't include `"0.0.0.0"`.

**Comment by Paul-calrec** [comment]
Thanks, Gareth, I'll fix that.

Paul

**Comment by Paul-calrec** [comment]
Well that has stopped it crashing. It is now returning
{
    "code": 400,
    "error": "Bad Request",
    "debug": "schema validation failed at /transport_params/0/interface_ip - no subschema has succeeded, but one of them is required to validate JSON - \"0.0.0.0\""
}

so presumably we are rejecting the SDP as being illegal

**Comment by Paul-calrec** [comment]
well I suppose it is up to me whether to accept 0.0.0.0 by allowing it in the constraints

**Comment by garethsb** [comment]
Yes, that's right. As you say, you can add that to your allowed values, or it may be that there's a way for you to map 0.0.0.0 to setting `rtp_enabled` to `false` on that leg...

**Comment by garethsb** [comment]
Yep, looks like you could wrap `nmos::parse_rtp_transport_file` and change the default behaviour to just ignore 0.0.0.0 or even to change `rtp_enabled`.

**Comment by Paul-calrec** [comment]
I'll have a look at that.
thanks again for your help Gareth

**Comment by David-P-B** [comment]
Hi, I don't know if this is releavant, but I've seen a similar crashing problem when I set a constraint for the receiver destination_port. e.g.
connection_receiver.data[nmos::fields::endpoint_constraints][0][nmos::fields::destination_port] = 7004;
If the receiver dst_port and sender dst_port do not match, then the receiver node crashes.

**Comment by garethsb** [comment]
That's not what a constraint should look like, that's why it crashes, same
problem as Paul.

**Comment by garethsb** [comment]
This line in the **pboettch/json-schema-validator** library is why we get a nasty SIGSEGV rather than a properly reported error or C++ exception (which would get mapped to an HTTP error status)... 

https://github.com/sony/nmos-cpp/blob/9577343d53fd31d5e05e5e183f818fa5904f8492/Development/third_party/nlohmann/json-validator.cpp#L1085

Should be improvable!

**Comment by garethsb** [comment]
Looks like this was added, see https://github.com/pboettch/json-schema-validator/commit/b28c15adea57bc9556cb00a7611536b5899dc67c

Let's try updating to more recent version.

**Comment by lo-simon** [comment]
Thanks Gareth, will test out the latest json-schema-validator.

**Comment by jonathan-r-thorpe** [comment]
Fixed by cba288e
